File: modules/common-net/src/main/java/com/shterneregen/securelan/common/net/transport/LengthPrefixedFrameChannel.java
package com.shterneregen.securelan.common.net.transport;

import java.io.Closeable;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.io.IOException;
import java.net.Socket;
import java.util.Objects;

public final class LengthPrefixedFrameChannel implements Closeable {
    public static final int DEFAULT_MAX_FRAME_SIZE_BYTES = 128 * 1024 * 1024;

    private final Socket socket;
    private final DataInputStream input;
    private final DataOutputStream output;
    private final int maxFrameSizeBytes;

    public LengthPrefixedFrameChannel(Socket socket) throws IOException {
        this(socket, DEFAULT_MAX_FRAME_SIZE_BYTES);
    }

    public LengthPrefixedFrameChannel(Socket socket, int maxFrameSizeBytes) throws IOException {
        this.socket = Objects.requireNonNull(socket, "socket");
        if (maxFrameSizeBytes < 1) {
            throw new IllegalArgumentException("maxFrameSizeBytes must be positive");
        }
        this.maxFrameSizeBytes = maxFrameSizeBytes;
        this.input = new DataInputStream(socket.getInputStream());
        this.output = new DataOutputStream(socket.getOutputStream());
    }

    public void writeUtf(String value) throws IOException {
        Objects.requireNonNull(value, "value");
        synchronized (output) {
            output.writeUTF(value);
            output.flush();
        }
    }

    public String readUtf() throws IOException {
        return input.readUTF();
    }

    public void writeFrame(byte[] bytes) throws IOException {
        Objects.requireNonNull(bytes, "bytes");
        if (bytes.length > maxFrameSizeBytes) {
            throw new IOException("Frame exceeds maximum size: " + bytes.length + " > " + maxFrameSizeBytes);
        }
        synchronized (output) {
            output.writeInt(bytes.length);
            output.write(bytes);
            output.flush();
        }
    }

    public byte[] readFrame() throws IOException {
        int length = input.readInt();
        if (length < 0) {
            throw new IOException("Negative frame length");
        }
        if (length > maxFrameSizeBytes) {
            throw new IOException("Frame exceeds maximum size: " + length + " > " + maxFrameSizeBytes);
        }
        byte[] bytes = input.readNBytes(length);
        if (bytes.length != length) {
            throw new IOException("Unexpected end of stream");
        }
        return bytes;
    }

    public String remoteAddress() {
        return socket.getRemoteSocketAddress().toString();
    }

    @Override
    public void close() throws IOException {
        socket.close();
    }
}

File: modules/common-net/src/main/kotlin/com/shterneregen/securelan/common/net/transport/LengthPrefixedFrameChannel.kt
package com.shterneregen.securelan.common.net.transport

import java.io.Closeable
import java.io.DataInputStream
import java.io.DataOutputStream
import java.io.IOException
import java.net.Socket
import java.util.Objects

class LengthPrefixedFrameChannel @Throws(IOException::class) constructor(
    private val socket: Socket,
    private val maxFrameSizeBytes: Int,
) : Closeable {
    private val input: DataInputStream
    private val output: DataOutputStream

    @Throws(IOException::class)
    constructor(socket: Socket) : this(socket, DEFAULT_MAX_FRAME_SIZE_BYTES)

    init {
        Objects.requireNonNull(socket, "socket")
        require(maxFrameSizeBytes >= 1) { "maxFrameSizeBytes must be positive" }
        input = DataInputStream(socket.getInputStream())
        output = DataOutputStream(socket.getOutputStream())
    }

    @Throws(IOException::class)
    fun writeUtf(value: String) {
        Objects.requireNonNull(value, "value")
        synchronized(output) {
            output.writeUTF(value)
            output.flush()
        }
    }

    @Throws(IOException::class)
    fun readUtf(): String = input.readUTF()

    @Throws(IOException::class)
    fun writeFrame(bytes: ByteArray) {
        Objects.requireNonNull(bytes, "bytes")
        if (bytes.size > maxFrameSizeBytes) {
            throw IOException("Frame exceeds maximum size: ${bytes.size} > $maxFrameSizeBytes")
        }
        synchronized(output) {
            output.writeInt(bytes.size)
            output.write(bytes)
            output.flush()
        }
    }

    @Throws(IOException::class)
    fun readFrame(): ByteArray {
        val length = input.readInt()
        if (length < 0) {
            throw IOException("Negative frame length")
        }
        if (length > maxFrameSizeBytes) {
            throw IOException("Frame exceeds maximum size: $length > $maxFrameSizeBytes")
        }
        val bytes = input.readNBytes(length)
        if (bytes.size != length) {
            throw IOException("Unexpected end of stream")
        }
        return bytes
    }

    fun remoteAddress(): String = socket.remoteSocketAddress.toString()

    @Throws(IOException::class)
    override fun close() {
        socket.close()
    }

    companion object {
        const val DEFAULT_MAX_FRAME_SIZE_BYTES: Int = 128 * 1024 * 1024
    }
}
